packages/lodestar/src/sync/range/chain.ts

```diff
@@ -0,0 +1,350 @@
+  startEpoch: Epoch;
+  /** A multi-threaded, non-blocking processor for applying messages to the beacon chain. */
+  private processChainSegment: ProcessChainSegment;
+  private downloadBeaconBlocksByRange: DownloadBeaconBlocksByRange;
+  private getPeersAndTargetEpoch: GetPeersAndTargetEpoch;
+  /** AsyncIterable that guarantees processChainSegment is run only at once at anytime */
+  private batchProcessor = new ItTrigger();
+  private maybeStuckTimeout!: NodeJS.Timeout; // clearTimeout(undefined) is okay
+  /** Sorted map of batches undergoing some kind of processing. */
+  private batches: Map<Epoch, Batch> = new Map();
+
+  /** Dynamic targetEpoch with associated peers. May be `null`ed if no suitable peer set exists */
+  private timeSeries = new TimeSeries({maxPoints: 1000});
+  private logger: ILogger;
+  private config: IBeaconConfig;
+  private signal: AbortSignal;
+  private opts: SyncChainOpts;
+
+  constructor(
+    localFinalizedEpoch: Epoch,
+    processChainSegment: ProcessChainSegment,
+    downloadBeaconBlocksByRange: DownloadBeaconBlocksByRange,
+    getPeersAndTargetEpoch: GetPeersAndTargetEpoch,
+    config: IBeaconConfig,
+    logger: ILogger,
+    signal: AbortSignal,
+    opts?: SyncChainOpts
+  ) {
+    this.startEpoch = localFinalizedEpoch;
+    this.processChainSegment = processChainSegment;
+    this.downloadBeaconBlocksByRange = downloadBeaconBlocksByRange;
+    this.getPeersAndTargetEpoch = getPeersAndTargetEpoch;
+    this.config = config;
+    this.logger = logger;
+    this.signal = signal;
+    this.opts = {
+      epochsPerBatch: opts?.epochsPerBatch ?? EPOCHS_PER_BATCH,
+      maybeStuckTimeoutMs: opts?.maybeStuckTimeoutMs ?? MAYBE_STUCK_TIMEOUT,
+    };
+
+    this.signal.addEventListener("abort", () => {
+      clearTimeout(this.maybeStuckTimeout);
+      this.batchProcessor.end(new ErrorAborted("SyncChain"));
+    });
+  }
+
+  /**
+   * For debugging and inspect the current status of active batches
+   */
+  batchesStatus(): BatchMetadata[] {
+    return toArr(this.batches).map((batch) => batch.getMetadata());
+  }
+
+  /**
+   * Main Promise that handles the sync process. Will resolve when initial sync completes
+   * i.e. when it successfully processes a epoch >= than this chain `targetEpoch`
+   */
+  async sync(): Promise<void> {
+    this.triggerBatchDownloader();
+    this.triggerBatchProcessor();
+
+    // Start processing batches on demand in strict sequence
+    // TODO: Use abortableSource
+    for await (const _ of this.batchProcessor) {
+      clearTimeout(this.maybeStuckTimeout);
+      this.maybeStuckTimeout = setTimeout(this.syncMaybeStuck, this.opts.maybeStuckTimeoutMs);
+
+      // Pull the current targetEpoch
+      const {targetEpoch} = this.getPeersAndTargetEpoch() || {};
+      if (targetEpoch === undefined) {
+        this.logger.verbose("No targetEpoch available");
+        continue;
+      }
+
+      // TEMP
+      validateBatchesStatus(toArr(this.batches));
```

Comment:
I guess this is needed for development, do we still need it for production?

Reply:
I'm not sure, I think we should keep testing and see if invalid situations happen or not.